lib/attrsets.nix

```diff
@@ -333,6 +333,39 @@ rec {
       ) (attrNames set)
     );
 
+   /*
```

Comment:
`foldAttrs` should have been a list fold over the attribute list, instead of a specific usage of that fold, and I don't think the specific usage chosen is particularly natural either. Very unfortunate.

However, calling this function `reduce` doesn't feel quite right either. It isn't present in Haskell, and its definition varies by language.
 - Scala: no initial value, like Haskell's `fold1 :: (a -> a -> a) -> list a -> a`
 - Python: no initial value either
 - JavaScript: optional initial value
 - MapReduce: pretty much disqualified for abuse of terminology, but also no initial value

I think we could call this `foldlAttrs` instead? That's more consistent with the existing Haskell-inspired naming, and could be complemented by `foldrAttrs` if we like. Sticking to patterns is important, because without consistency, it's harder to guess the meaning of a function, requiring more memorization and more searching for the right function. At some point that defeats the benefit of defining the extra functions. See [Fairbairn threshold](https://wiki.haskell.org/Fairbairn_threshold).

-----------

_some more details_

In Haskell, `fold` folds by applying the `Monoid` typeclass, which is a merging operation accompanying the type of the items; similar to how we merge option value definitions in the module system. In this regard, the "abuse" of the `foldAttrs` name is probably not so bad. I don't think we badly need to merge attribute values according to an option type.

```haskell
-- Recurse into the `tail`, which is on the right.
-- Outermost call processes the `head`, which for attrsets would be the lowest key, along with its value.
foldr :: (a -> b -> b) -> b -> list a -> b
-- Recurse into the `init`, which is on the left.
-- Outermost call processes the `last`, which for us is the highest key.
foldl :: (b -> a -> b) -> b -> list a -> b
```